// nodescore/build/lib/player.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _util = require('./util');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/** Class representing a player. */
var Player =
/**
 * creates a **Player** object
 *
 * @param {NodeScore} wrapper - the API wrapper object
 * @param {Object} player - team from the JSON
 */
function Player(wrapper, player) {
    _classCallCheck(this, Player);

    var p = player;
    this.id = parseInt(p.id);
    this.firstName = p.firstname;
    this.lastName = p.lastname;
    this.age = parseInt(p.age);
    this.height = p.height == null ? null : parseHeight(p.height);
    this.weight = p.weight == null ? null : parseWeight(p.weight);
    this.birthDate = (0, _util.createDateObject)(p.birthdate);
    this.birthPlace = p.birthplace;
    this.birthCounrty = p.birthcountry;
    this.position = p.position;
    this.teamId = parseInt(p.teamid);
    this.teamName = p.team;
    this.nationality = p.nationality;
};

/**
 * creates height integer from string from JSON
 *
 * @param {string} heightString - the heightstring from the JSON
 * @return {number} - the int created from the heightstring
 */


exports.default = Player;
function parseHeight(heightString) {
    var ss = heightString.split(' cm');
    return parseInt(ss);
}

/**
 * creates weight integer from string from JSON
 *
 * @param {string} weightString - the weightstring from the JSON
 * @return {number} - the int created from the weightstring
 */
function parseWeight(weightString) {
    var ss = weightString.split(' kg');
    return parseInt(ss);
}
